packages/core/src/storage.ts

```diff
@@ -136,28 +168,171 @@ export function describeError(e: StorageOperation): string {
-    } else if (nameMatches) {
-      return 'typechange';
+  const originalDetailed = getDetailedLayout(original);
+  const updatedDetailed = getDetailedLayout(updated);
+  return getStorageUpgradeErrorsGeneric(originalDetailed, updatedDetailed, { canGrow: true });
+}
+
+interface StorageField {
+  label: string;
+  type: ParsedTypeDetailed;
+}
+
+// This function is generic because it can compare top level storage layout as well as struct layout.
+function getStorageUpgradeErrorsGeneric<T extends StorageField>(
+  original: T[],
+  updated: T[],
+  { canGrow }: { canGrow: boolean },
+): StorageOperation<T>[] {
+  const ops = levenshtein(original, updated, (a, b) => matchStorageField(a, b, { canGrow: false }));
+  if (canGrow) {
+    // appending is not an error in this case
+    return ops.filter(o => o.kind !== 'append');
+  } else {
+    return ops;
+  }
+}
+
+type Replace<T, K extends string, V> = Omit<T, K> & Record<K, V>;
+
+function getDetailedLayout(layout: StorageLayout): StorageItemDetailed[] {
+  function parseWithDetails<I extends { type: string }>(item: I): Replace<I, 'type', ParsedTypeDetailed> {
+    return {
+      ...item,
+      type: addDetailsToParsedType(parseTypeId(item.type)),
+    };
+  }
+
+  function addDetailsToParsedType(parsed: ParsedTypeId): ParsedTypeDetailed {
+    const item = layout.types[parsed.id];
+    const members =
+      item?.members && (isStructMembers(item?.members) ? item.members.map(parseWithDetails) : item?.members);
+    return {
+      ...parsed,
+      args: parsed.args?.map(addDetailsToParsedType),
+      rets: parsed.args?.map(addDetailsToParsedType),
+      item: { ...item, members },
+    };
+  }
+
+  return layout.storage.map(parseWithDetails);
+}
+
+function isEnumMembers<T>(members: TypeItemMembers<T>): members is EnumMember[] {
+  return members.length === 0 || typeof members[0] === 'string';
+}
+
+function isStructMembers<T>(members: TypeItemMembers<T>): members is StructMember<T>[] {
+  return members.length === 0 || typeof members[0] === 'object';
+}
+
+function matchStorageField(original: StorageField, updated: StorageField, { canGrow }: { canGrow: boolean }) {
+  const nameMatches = original.label === updated.label;
+  const typeMatches = compatibleTypes(original.type, updated.type, { canGrow });
+
+  if (typeMatches && nameMatches) {
+    return 'equal';
+  } else if (typeMatches) {
+    return 'rename';
+  } else if (nameMatches) {
+    return 'typechange';
+  } else {
+    return 'replace';
+  }
+}
+
+// TODO: Avoid getting stuck with recursive types.
+// struct S { mapping (uint => S) s; }
+
+function compatibleTypes(
+  original: ParsedTypeDetailed,
+  updated: ParsedTypeDetailed,
+  { canGrow }: { canGrow: boolean },
+): boolean {
+  if (original.head !== updated.head) {
+    return false;
+  }
+
+  if (original.args === undefined || updated.args === undefined) {
+    // both should be undefined at the same time
+    assert(original.args === updated.args);
```

Comment:
Can we add a message to each assertion? It will help debug user errors in the future.